hal/stm32f1.c

```diff
@@ -0,0 +1,342 @@
+#define RCC_CFGR_PPRE1_DIV_2        (4 << 8)
+#define RCC_CFGR_PPRE1_DIV_4        (5 << 8)
+
+#define RCC_CFGR_PPRE2_MASK         0x3800
+#define RCC_CFGR_PPRE2_DIV_NONE     (0 << 11)
+#define RCC_CFGR_PPRE2_DIV_2        (4 << 11)
+#define RCC_CFGR_PPRE2_DIV_4        (5 << 11)
+
+#define PLL_FULL_MASK (0x003F0000)
+#define RCC_CFGR_PLLSRC             (1 << 22)
+#define RCC_CFGR_PLLMUL_MUL_9       (7 << 18)
+
+/*** FLASH ***/
+#define RCC_APB1ENR         (*(volatile uint32_t *)(RCC_BASE + 0x1C))
+#define RCC_APB1ENR_PWREN   (1 << 28)
+
+#define FLASH_BASE          (0x40022000U)
+#define FLASH_ACR           (*(volatile uint32_t *)(FLASH_BASE + 0x00U))
+#define FLASH_KEYR          (*(volatile uint32_t *)(FLASH_BASE + 0x04U))
+#define FLASH_SR            (*(volatile uint32_t *)(FLASH_BASE + 0x0CU))
+#define FLASH_CR            (*(volatile uint32_t *)(FLASH_BASE + 0x10U))
+#define FLASH_AR            (*(volatile uint32_t *)(FLASH_BASE + 0x14U))
+
+#define FLASH_MAX_SZ        (128*1024) /* only low- and medium-density devices supported */
+#define FLASH_PAGE_SZ       (1024)
+
+/* Register values */
+#define FLASH_ACR_ENABLE_PRFT                 (1 << 4)
+
+#define FLASH_SR_BSY                          (1 << 0)
+
+#define FLASH_CR_LOCK                         (1 << 7)
+#define FLASH_CR_STRT                         (1 << 6)
+#define FLASH_CR_PER                          (1 << 1)
+#define FLASH_CR_PG                           (1 << 0)
+
+#define FLASH_KEY1                            (0x45670123U)
+#define FLASH_KEY2                            (0xCDEF89ABU)
+
+static void RAMFUNCTION flash_set_waitstates(int waitstates)
+{
+    FLASH_ACR |=  waitstates | FLASH_ACR_ENABLE_PRFT;
+}
+
+static int RAMFUNCTION valid_flash_area(uint32_t address, int len)
+{
+    if (len <= 0 || len > FLASH_MAX_SZ)
+        return -1;
+    if (address < ARCH_FLASH_OFFSET || address >= (ARCH_FLASH_OFFSET + FLASH_MAX_SZ))
+        return -1;
+    if ((address + len) > (ARCH_FLASH_OFFSET + FLASH_MAX_SZ))
+        return -1;
+
+    return 0;
+}
+
+static inline void RAMFUNCTION flash_wait_complete(void)
+{
+    while ((FLASH_SR & FLASH_SR_BSY) == FLASH_SR_BSY)
+        ;
+}
+
+static int RAMFUNCTION flash_erase_page(uint32_t address)
+{
+    const uint32_t end = address + FLASH_PAGE_SZ;
+
+    flash_wait_complete();
+
+    FLASH_CR |= FLASH_CR_PER;
+    FLASH_AR = address;
+    FLASH_CR |= FLASH_CR_STRT;
+    flash_wait_complete();
+    FLASH_CR &= ~FLASH_CR_PER;
+
+    /* verify erase */
+    while (address < end) {
+        if (*(uint32_t*)address != (uint32_t)~0)
+            return -1;
+
+        address += 4;
+    }
+
+    return 0;
+}
+
+int RAMFUNCTION flash_w16(volatile uint16_t *dst, const uint16_t value)
+{
+    /* do the write */
+    *dst = value;
+    DMB();
+    flash_wait_complete();
+
+    /* verify */
+    return (*dst == value) ? 0 : -1;
+}
+
+int RAMFUNCTION hal_flash_write(uint32_t address, const uint8_t *data, int len)
+{
```

Comment:
nit: please reduce the number of empty lines in this function.

Reply:
Done!